projects/kit/components/slider/slider-key-steps.directive.ts

```diff
@@ -0,0 +1,139 @@
+import {
+    ChangeDetectorRef,
+    Directive,
+    ElementRef,
+    forwardRef,
+    HostListener,
+    Inject,
+    Input,
+    Optional,
+    Self,
+} from '@angular/core';
+import {NgControl} from '@angular/forms';
+import {
+    AbstractTuiControl,
+    isNativeFocused,
+    TUI_FOCUSABLE_ITEM_ACCESSOR,
+    tuiDefaultProp,
+    TuiFocusableElementAccessor,
+} from '@taiga-ui/cdk';
+import {TuiKeySteps} from '@taiga-ui/kit/types';
+
+import {TuiSliderComponent} from './slider.component';
+
+/**
+ * TODO: remove in ivy compilation
+ * @deprecated DONT USE IT (for internal purpose only). Can be deleted at any time!
+ */
+export function checkHasMinMaxPercents(steps: TuiKeySteps): boolean {
+    return !steps.length || (steps[0][0] === 0 && steps[steps.length - 1][0] === 100);
+}
+
+// @dynamic
+@Directive({
+    selector: 'input[tuiSlider][keySteps]',
+    providers: [
+        {
+            provide: TUI_FOCUSABLE_ITEM_ACCESSOR,
+            useExisting: forwardRef(() => TuiSliderComponent),
+        },
+    ],
+})
+export class TuiSliderKeyStepsDirective
+    extends AbstractTuiControl<number>
+    implements TuiFocusableElementAccessor
+{
+    @Input()
+    @tuiDefaultProp(checkHasMinMaxPercents, 'Should contain min and max values')
+    keySteps: TuiKeySteps = [];
+
+    get nativeFocusableElement(): HTMLInputElement | null {
+        return this.computedDisabled ? null : this.elementRef.nativeElement;
+    }
+
+    get focused(): boolean {
+        return isNativeFocused(this.nativeFocusableElement);
+    }
+
+    constructor(
+        @Optional()
+        @Self()
+        @Inject(NgControl)
+        control: NgControl | null,
+        @Inject(ChangeDetectorRef) changeDetectorRef: ChangeDetectorRef,
+        @Inject(ElementRef) private readonly elementRef: ElementRef<HTMLInputElement>,
+        @Inject(TuiSliderComponent) private readonly slider: TuiSliderComponent,
+    ) {
+        super(control, changeDetectorRef);
+    }
+
+    @HostListener('change')
+    onInput() {
+        const {valuePercentage} = this.slider;
+        const [lowerStep, upperStep] = this.findKeyStepsBoundariesByFn(
+            ([keyStepPercentage, _]) => valuePercentage <= keyStepPercentage,
+        );
+        const newControlValue = this.transformToControlValue(
+            valuePercentage,
+            lowerStep,
+            upperStep,
+        );
+
+        this.updateValue(newControlValue);
+    }
+
+    writeValue(controlValue: number | null) {
+        if (controlValue === null) {
+            return;
+        }
+
+        const [lowerStep, upperStep] = this.findKeyStepsBoundariesByFn(
+            ([_, keyStepValue]) => controlValue <= keyStepValue,
+        );
+
+        this.slider.value = this.transformToNativeValue(
+            controlValue,
+            lowerStep,
+            upperStep,
+        );
+    }
+
+    protected getFallbackValue(): number {
+        return 0;
+    }
+
+    private findKeyStepsBoundariesByFn(
```

Comment:
Let's move to a separate utility function since it's pure. Might even be non-exported